<font color = green >

# <center>Homework<center>
</font>

__Exploratory Data Analysis (EDA)__ of car insurance dataset.
    
__Dataset description:__ Car insurance claim. As you are working in the insurance company. Company wants to know the reason why claim was not made. Doing so would allow insurance company to improve there policy for giving loan to the customer. In this project you are dealing with various feature such as age, occupation etc. based on that let's get back to the final conculsion.    

__Steps:__
- Data Cleaning
- Descriptive Statistics
- Data Visualization
- Conclusions

In [125]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

df = pd.read_csv("../data/Car_Insurance_Claim.csv", index_col='ID')
df.drop(columns='Unnamed: 0', inplace=True)
df.columns = df.columns.str.lower()
df.rename(columns={'duis': 'dui_violations'}, inplace=True) # dui = driving under the influence
df.head()

age  gender      race driving_experience    education  \
ID                                                                
569520    65+  female  majority               0-9y  high school   
750365  16-25    male  majority               0-9y         none   
199901  16-25  female  majority               0-9y  high school   
478866  16-25    male  majority               0-9y   university   
731664  26-39    male  majority             10-19y         none   

               income        credit_score vehicle_ownership vehicle_year  \
ID                                                                         
569520    upper class   0.629027313918201               1.0   after 2015   
750365        poverty  0.3577571170184627               0.0  before 2015   
199901  working class  0.4931457852181982               1.0  before 2015   
478866  working class  0.2060128507324556               1.0  before 2015   
731664  working class  0.3883658881572184               1.0  before 2015   

       married children postal_code annual_mileage vehicle_type  \
ID                                                                
569520     0.0      1.0       10238        12000.0        sedan   
750365     0.0      0.0       10238        16000.0        sedan   
199901     0.0      0.0       10238        11000.0        sedan   
478866     0.0     none       32765        11000.0        sedan   
731664     0.0      0.0     32765.0        12000.0        sedan   

       speeding_violations dui_violations past_accidents outcome  
ID                                                                
569520                   0              0              0     0.0  
750365                   0              0              0     1.0  
199901                   0              0              0     0.0  
478866                   0              0              0     0.0  
731664                   2              0              1     1.0

<p style="color:red;">Disclaimer: this notebook contains a lot of thinking process, presented here in educational purposes only. Main conclusions are at the end.</p>

# **Data Cleaning**

In [126]:
df.info()
df.index.unique()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10000 entries, 569520 to 442696
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   age                  9950 non-null   object
 1   gender               9950 non-null   object
 2   race                 9951 non-null   object
 3   driving_experience   9950 non-null   object
 4   education            9951 non-null   object
 5   income               10000 non-null  object
 6   credit_score         8978 non-null   object
 7   vehicle_ownership    9950 non-null   object
 8   vehicle_year         9950 non-null   object
 9   married              9952 non-null   object
 10  children             9950 non-null   object
 11  postal_code          9950 non-null   object
 12  annual_mileage       9000 non-null   object
 13  vehicle_type         9951 non-null   object
 14  speeding_violations  9950 non-null   object
 15  dui_violations       9950 non-null   object
 16

Int64Index([569520, 750365, 199901, 478866, 731664, 877557, 930134, 461006,
             68366, 445911,
            ...
            302403,  67445, 188300, 156495, 752887, 323164, 910346, 468409,
            903459, 442696],
           dtype='int64', name='ID', length=10000)

- Dataset contains a total of 10k entries, and the ID column contains 10k unique entries, so it is a valid index. Therefore the Unnamed column was dropped correctly.
- All the columns are objects, either because of NaNs or because all the data is stored as strings.
- Only the INCOME column has no null values, but it is still object, so I assume that all values in the dataset are strings.
- Most of the columns have exactly 9950 or 9951 non-null values. Maybe there are 50 empty rows?
- CREDIT_SCORE is the column with the biggest number of null values.

In [127]:
df[df['age'].isna()]

age  gender      race driving_experience    education         income  \
ID                                                                             
22475   NaN  female  majority               0-9y  high school   middle class   
55182   NaN  female  majority             10-19y  high school  working class   
626735  NaN  female  majority               0-9y   university  working class   
420204  NaN    male  majority             20-29y   university    upper class   
45019   NaN  female  majority               0-9y   university   middle class   
50734   NaN  female  majority               0-9y         none        poverty   
67899   NaN    male  majority               30y+  high school    upper class   
183837  NaN  female  majority             20-29y   university    upper class   
976721  NaN  female  majority             20-29y   university    upper class   
513565  NaN  female  majority               0-9y  high school  working class   
110283  NaN  female  majority               0-9y   university    upper class   
847188  NaN  female  majority             20-29y   university   middle class   
883171  NaN  female  minority               0-9y  high school        poverty   
384342  NaN  female  majority             20-29y   university    upper class   
872748  NaN    male  majority               0-9y   university    upper class   
172866  NaN  female  majority               0-9y         none  working class   
4747    NaN  female  majority             20-29y         none   middle class   
314564  NaN  female  majority               0-9y         none        poverty   
209049  NaN    male  majority             10-19y   university    upper class   
57078   NaN    male  majority             20-29y   university    upper class   
58611   NaN  female  majority               0-9y  high school        poverty   
141111  NaN    male  majority               0-9y  high school   middle class   
634309  NaN    male  majority             10-19y         none  working class   
352133  NaN    male  majority               30y+   university    upper class   
135855  NaN    male  majority               0-9y   university    upper class   
915389  NaN  female  majority             10-19y         none        poverty   
292609  NaN  female  majority             10-19y   university    upper class   
403549  NaN  female  minority             10-19y   university   middle class   
509738  NaN    male  majority               0-9y   university    upper class   
328633  NaN    male  majority               0-9y   university   middle class   
703183  NaN  female  majority             20-29y  high school    upper class   
346384  NaN  female  majority             20-29y   university    upper class   
862136  NaN    male  majority             10-19y   university  working class   
839283  NaN    male  majority             20-29y   university    upper class   
624831  NaN    male  minority             20-29y  high school    upper class   
161972  NaN    none  majority               0-9y   university    upper class   
278061  NaN    male  minority             10-19y  high school   middle class   
343615  NaN    male  majority               30y+   university    upper class   
143667  NaN  female  majority             20-29y  high school  working class   
481006  NaN    male  majority             10-19y  high school    upper class   
618914  NaN  female      none             10-19y   university    upper class   
896224  NaN  female  minority               30y+  high school    upper class   
319786  NaN    male  majority               0-9y   university    upper class   
740411  NaN    male  majority             10-19y   university    upper class   
841092  NaN    male  majority             10-19y  high school   middle class   
930407  NaN    male  majority             20-29y  high school    upper class   
296999  NaN  female  majority             20-29y  high school  working class   
542940  NaN    male  majority             10-19y  high school    upper class   
385216  

- So 50 NA's exactly in almost every column is likely just a coincidence.

In [128]:
df.describe()

age  gender      race driving_experience    education       income  \
count    9950    9950      9951               9950         9951        10000   
unique      5       3         3                  5            3            4   
top     26-39  female  majority               0-9y  high school  upper class   
freq     3035    4968      8919               3503         4116         4336   

       credit_score vehicle_ownership vehicle_year married children  \
count          8978              9950         9950    9952     9950   
unique         8929                 3            3       3        3   
top            none               1.0  before 2015     0.0      1.0   
freq             50              6907         6896    4972     6815   

       postal_code annual_mileage vehicle_type speeding_violations  \
count         9950           9000         9951                9950   
unique           9             22            3                  22   
top          10238        11000.0        sedan                   0   
freq          6223           1244         9432                4978   

       dui_violations past_accidents outcome  
count            9950           9950    9950  
unique              8             16       3  
top                 0              0     0.0  
freq             8041           5528    6795

- Postal Code column has only 9 unique values wich may possibly mean that the data has been gathered in one location only, and that will be the problem if need to make conclusion for a bigger area or our model trained on this data is supposed to work on bigger area.
- Annual mileage also has not so many unique values which is suspicios. Maybe the mileage was rounded? Or the data was gathered with low precision?

Let's check all the columns for null values that are not NaN and for values that look like null but are actually a category. In the first case we want to replace them with NaN and in the second case we might want to rename them.

In [129]:
noncat_columns = ['credit_score', 'postal_code', 'annual_mileage', 'speeding_violations', 'dui_violations', 'past_accidents']
pd.crosstab(**df.iloc[:, ~df.columns.isin(noncat_columns)].melt(var_name='columns', value_name='index'))

columns         age  children  driving_experience  education  gender  income  \
index                                                                          
0-9y              0         0                3503          0       0       0   
0.0               0      3085                   0          0       0       0   
1.0               0      6815                   0          0       0       0   
10-19y            0         0                3264          0       0       0   
16-25          1995         0                   0          0       0       0   
20-29y            0         0                2093          0       0       0   
26-39          3035         0                   0          0       0       0   
30y+              0         0                1040          0       0       0   
40-64          2907         0                   0          0       0       0   
65+            1963         0                   0          0       0       0   
after 2015        0         0                   0          0       0       0   
before 2015       0         0                   0          0       0       0   
female            0         0                   0          0    4968       0   
high school       0         0                   0       4116       0       0   
majority          0         0                   0          0       0       0   
male              0         0                   0          0    4933       0   
middle class      0         0                   0          0       0    2138   
minority          0         0                   0          0       0       0   
none             50        50                  50       1947      49       0   
poverty           0         0                   0          0       0    1814   
sedan             0         0                   0          0       0       0   
sports car        0         0                   0          0       0       0   
university        0         0                   0       3888       0       0   
upper class       0         0                   0          0       0    4336   
working class     0         0                   0          0       0    1712   

columns        married  outcome  race  vehicle_ownership  vehicle_type  \
index                                                                    
0-9y                 0        0     0                  0             0   
0.0               4972     6795     0               2993             0   
1.0               4930     3105     0               6907             0   
10-19y               0        0     0                  0             0   
16-25                0        0     0                  0             0   
20-29y               0        0     0                  0             0   
26-39                0        0     0                  0             0   
30y+                 0        0     0                  0             0   
40-64                0        0     0                  0             0   
65+                  0        0     0                  0             0   
after 2015           0        0     0                  0             0   
before 2015          0        0     0                  0             0   
female               0        0     0                  0             0   
high school          0        0     0                  0             0   
majority             0        0  8919                  0             0   
male                 0        0     0                  0             0   
middle class         0        0     0                  0             0   
minority             0        0   982                  0             0   
none                50       50    50                 50            50   
poverty              0        0     0                  0             0   
sedan                0        0     0                  0          9432   
sports car           0        0     0                  0           469   
university           0        0     0                  0       

'none' value is really popular here.
- It seems that 'none' value ought to be replaced with NaN as their meaning is NA in listed columns: Age, Children, Driving Experience, Gender, Married, Outcome, Race, Vehicle Ownership, Vehicle Year.
- Although, for column Vehicle Type we should consider changing 'none' to 'Other' and in column Education 'none' can mean 'No education'. And there are actualy 1947 'none' values in Education so I'm sure about that one.

But first, I'd like to check why so many columns have 50 NAs exactly.

In [130]:
df[df['age'] == 'none'].apply(lambda x: x[x == 'none'].count())

age                    50
gender                  1
race                    0
driving_experience      0
education              12
income                  0
credit_score            0
vehicle_ownership       0
vehicle_year            0
married                 0
children                0
postal_code             1
annual_mileage          0
vehicle_type            0
speeding_violations     1
dui_violations          1
past_accidents          0
outcome                 0
dtype: int64

In [131]:
df[df['children'] == 'none'].apply(lambda x: x[x == 'none'].count())

age                     0
gender                  0
race                    0
driving_experience      0
education              15
income                  0
credit_score            0
vehicle_ownership       1
vehicle_year            0
married                 0
children               50
postal_code             0
annual_mileage          0
vehicle_type            0
speeding_violations     0
dui_violations          1
past_accidents          0
outcome                 0
dtype: int64

In [132]:
df[df['outcome'] == 'none'].apply(lambda x: x[x == 'none'].count())

age                     0
gender                  0
race                    1
driving_experience      1
education              10
income                  0
credit_score            1
vehicle_ownership       0
vehicle_year            1
married                 0
children                0
postal_code             0
annual_mileage          0
vehicle_type            0
speeding_violations     0
dui_violations          0
past_accidents          0
outcome                50
dtype: int64

- My suspicion was false: these 50 NaNs are spread randomly and do not cross in same 50 rows. So I will just go on with them and change them to np.nan.

- Still, I don't see how records without Outcome can help in my analysis, so I will drop them.

- In columns that are not categorical and are not shown above, 'none' could mean 0, but since there are lot of actual 0's in each column, I assume that 'none' value here mean NA. Now only Annual Mileage should be also checked for zeros.

In [133]:
df['annual_mileage'].value_counts()

11000.0    1244
12000.0    1200
13000.0    1133
10000.0    1060
14000.0     881
9000.0      841
15000.0     626
8000.0      552
16000.0     416
7000.0      317
17000.0     242
6000.0      165
18000.0     103
5000.0       65
none         50
19000.0      48
4000.0       27
20000.0      13
3000.0       10
21000.0       3
2000.0        2
22000.0       2
Name: annual_mileage, dtype: int64

- And really, 'none' here migh have meant a 0 mileage. But since 3k and 2k values are so rare, I suppose that zero mileage wouldn't have much sence here. And since there are 50 'none' values they cannot mean 0s. Thus they should also be marked np.nan.

In [136]:
df.loc[:, df.columns != 'education'] = df.loc[:, df.columns != 'education'].replace('none', np.nan)
df['education'].replace('none', 'no degree', inplace=True)
df = df[~df['outcome'].isna()]
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9900 entries, 569520 to 442696
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   age                  9801 non-null   object
 1   gender               9801 non-null   object
 2   race                 9804 non-null   object
 3   driving_experience   9801 non-null   object
 4   education            9852 non-null   object
 5   income               9900 non-null   object
 6   credit_score         8839 non-null   object
 7   vehicle_ownership    9801 non-null   object
 8   vehicle_year         9801 non-null   object
 9   married              9803 non-null   object
 10  children             9802 non-null   object
 11  postal_code          9801 non-null   object
 12  annual_mileage       8859 non-null   object
 13  vehicle_type         9803 non-null   object
 14  speeding_violations  9801 non-null   object
 15  dui_violations       9801 non-null   object
 16 

Probably last step in cleaning data: changing type for numeric columns.

In [135]:
numeric_columns = ['credit_score', 'vehicle_ownership', 'married', 'children', 'annual_mileage', 'speeding_violations', 'dui_violations', 'past_accidents', 'outcome']

# **Descriptive Statistics**

# **Data Visualisation**

# **Conclusions**